library/src/SPSApplication.ts

```diff
@@ -1,97 +1,263 @@
+import { LoadingOverlay } from './LoadingOverlay';
+import { SPSSignalling } from './SignallingExtension';
+import { MessageStats } from './Messages';
+import { v4 as uuidv4 } from 'uuid';
+
+// For local testing. Declare a websocket URL that can be imported via a .env file that will override 
+// the signalling server URL builder.
+declare var WEBSOCKET_URL: string;
+declare var ENABLE_METRICS: boolean;
+
+const SupportedStats : Record<string, string> = {
+    'video_width': 'width',
+    'video_height': 'height',
+    'video_bitrate': 'bitrate',
+    'video_dropped': 'dropped',
+    'video_packets_lost': 'packets lost',
+    'video_fps': 'fps',
+    'video_pli_count': 'count',
+    'video_keyframes': 'keyframes',
+    'video_nack_count': 'nacks',
+    'video_freeze_count': 'freeze',
+    'video_jitter': 'jitter',
+    'video_frame_count': 'frame count',
+    'audio_bitrate': 'audio bitrate'
+}
+
+export class SPSApplication extends Application {
+	private loadingOverlay: LoadingOverlay;
+	private signallingExtension: SPSSignalling;
+    private statsTimer: ReturnType<typeof setInterval> | undefined;
+    private sessionData: any;
+    private loadingStart: number | undefined;
+
+	static Flags = class {
+		static sendToServer = "sendStatsToServer"
+	}
+
+	constructor(config: UIOptions) {
+		super(config);
+		this.signallingExtension = new SPSSignalling(this.stream.webSocketController);
+		this.signallingExtension.onAuthenticationResponse = this.handleSignallingResponse.bind(this);
+		this.signallingExtension.onInstanceStateChanged = this.handleSignallingResponse.bind(this);
+
+		this.enforceSpecialSignallingServerUrl();
+
+		// Add 'Send Stats to Server' checkbox
+		const spsSettingsSection = this.configUI.buildSectionWithHeading(this.settingsPanel.settingsContentElement, "Scalable Pixel Streaming");
+		const sendStatsToServerSetting = new SettingFlag(
+			SPSApplication.Flags.sendToServer,
+			"Send stats to server",
+			"Send session stats to the server",
+			false,
+			this.stream.config.useUrlParams
+		);
+
+		spsSettingsSection.appendChild(new SettingUIFlag(sendStatsToServerSetting).rootElement);
+		this.loadingOverlay = new LoadingOverlay(this.stream.videoElementParent);
+
+		this.stream.addEventListener(
+			'statsReceived',
+			({ data: { aggregatedStats } }) => {
+                this.onSessionStats(aggregatedStats);
+				if (sendStatsToServerSetting.flag) {
+					this.sendStatsToSignallingServer(aggregatedStats);
+				}
+			}
+		);
+
+        this.stream.addEventListener('webRtcConnected', () => this.startSession() );
+	}
+
+    startSession() {
+        if (!ENABLE_METRICS) {
+            return;
+        }
+
+        // generate a unique session id
+        const sessionId: string = uuidv4();
+
+        // register end of session event
+        window.addEventListener('beforeunload', () => this.endSession(sessionId));

```

Comment:
Does this capture the stats if the UE side or the SS side terminates? E.g. Websocket disconnect case? If not can you try to capture the case where the session ends but the browser has not navigated away.

Reply:
It might as simple as something like

`this.stream.addEventListener('webRtcDisconnected', () => this.endSession(sessionId) );`

Reply:
Additionally, can we add a special stat to track websocket disconnect (or better yet, to track session end reason)

Reply:
So that's similar to something I've been thinking about today. Things like loading time, user agent and this aren't reported currently because they're not ongoing gauge metrics like bit rate etc. They're more like events. Well this disconnect event is, but user agent and loading time are one time metrics. I'm not sure how they should be handled.
I know the buccaneer setup uses loki for event like stats that show up as annotations on the grafana graphs.. but is that how we want to handle each of these stats? As events?